### ISSUE TITLE
Conflict with Multiple File Upload field

### ISSUE THREAD

**Issue by Avakulenko** [tyxla/Gravity-Forms-Multiple-Form-Instances#20]
Gravity Forms: Multiple Form Instances version: 1.1.1
Wordpress version: 4.6.1

Plugin prevents "delete file"  event in the Multiple File Upload field.

**Comment by rrobinpro** [comment]
I have same problem with Multiple Form Instances version 1.1.1 and Wordpress version 4.6.1

**Comment by medfreeman** [comment]
You can solve this by adding the following snippet in your theme's functions.php file:

```
if ( ! function_exists( 'gform_multiple_upload_delete_file_fix' ) ) {
	function gform_multiple_upload_delete_file_fix( $strings ) {
		if ( isset( $strings['\'gf_2\''] ) ) {
			$form_id = str_replace ( 'gf_', '', $strings['\'gf_2\'']);
			$form_id = str_replace ( '\'', '', $form_id);
			$form_id = absint( $form_id );
			if ( $form_id ) {
				$strings["onclick='gformDeleteUploadedFile(2"] = "onclick='gformDeleteUploadedFile(" . $form_id;
				$strings["onkeypress='gformDeleteUploadedFile(2"] = "onkeypress='gformDeleteUploadedFile(" . $form_id;
			}
		}
		return $strings;
	}
}

add_filter( 'gform_multiple_instances_strings', 'gform_multiple_upload_delete_file_fix' );
```

As you can see, my example only covers form with ID 2.
This plugin's filter 'gform_multiple_instances_strings' should obviously pass the original form id and the new random form id, so these string manipulations are not needed.

I'll try to make a few pull requests when i got more time, one to extend the filter parameters, and the second to simply add the needed string for file uploads deletion in https://github.com/tyxla/Gravity-Forms-Multiple-Form-Instances/blob/master/gravityforms-multiple-form-instances.php

**Comment by rrobinpro** [comment]
Thanks medfreeman

Your code is correct when you reload the form but not live when you upload a document.

**Comment by medfreeman** [comment]
Update that works with multifile uploads:

```
if ( ! function_exists( 'gform_multiple_upload_delete_file_fix' ) ) {
	function gform_multiple_upload_delete_file_fix( $strings ) {
		if ( isset( $strings['\'gf_2\''] ) ) {
			$form_id = str_replace ( 'gf_', '', $strings['\'gf_2\'']);
			$form_id = str_replace ( '\'', '', $form_id);
			$form_id = absint( $form_id );
			if ( $form_id ) {
				$strings["onclick='gformDeleteUploadedFile(2"] = "onclick='gformDeleteUploadedFile(" . $form_id;
				$strings["onkeypress='gformDeleteUploadedFile(2"] = "onkeypress='gformDeleteUploadedFile(" . $form_id;
                                $strings['gform_multifile_upload_2_'] = 'gform_multifile_upload_' . $form_id . '_';
				$strings['gform_drag_drop_area_2_'] = 'gform_drag_drop_area_' . $form_id . '_';
				$strings['gform_browse_button_2_'] = 'gform_browse_button_' . $form_id . '_';
				$strings['gform_preview_2_'] = 'gform_preview_' . $form_id . '_';
				$strings['gform_multifile_messages_2_'] = 'gform_multifile_messages_' . $form_id . '_';
			}
		}
		return $strings;
	}
}

add_filter( 'gform_multiple_instances_strings', 'gform_multiple_upload_delete_file_fix' );
```

And you'll also need this javascript snippet somewhere:

```
var mgformDeleteUploadedFile = function(formid, fieldid, element) {
	var $this = $(element);
	var randomid = $this.parents('form').find('input:hidden[name=gform_random_id]').val();
	gformDeleteUploadedFile(randomid, fieldid, element);
};

gform.addFilter(
	'gform_file_upload_markup',
	function(html) {
            return html.replace('gformDeleteUploadedFile(2', 'mgformDeleteUploadedFile(2');
        }
);
```

**Comment by tyxla** [comment]
@medfreeman thank you for this! Would you like to contribute it as a PR so we can fix that in the next version?

**Comment by medfreeman** [comment]
Thanks.
I'll gladly submit a PR, but it seems i've spoken too fast.
This code results in the form not submitting the uploaded file at all.
I'm looking into it, and i'll correct the code when it's done.

**Comment by medfreeman** [comment]
I'm afraid the approach i used is the best i can come up with.
This implies having a part of a gravityforms js file in the source, with little modifications.
I'm not thrilled with it, as it could need maintenance, but the way gravity is coded now doesn't allow a better approach (do not hesitate to correct me if i'm wrong !).

**Comment by medfreeman** [comment]
I just figured that i can do this in a simpler way.
Hook on plupload 'beforeUpload' event, change multipart_params form_id to the original form id by taking the id stored in the field named 'gform_wrapper_original_id_'
That should work and only require a few lines of code, and nothing from gravity's source.
I'll close the actual PR and try to submit a new one this afternoon.

**Comment by tyxla** [comment]
@medfreeman that would be really awesome. I'm reluctant to introduce an entire file from a premium plugin in an OSS repo :) Thank you!

**Comment by medfreeman** [comment]
done!

**Comment by medfreeman** [comment]
there's a few spacing problems, i'll re-send one.

**Comment by tyxla** [comment]
Thanks @medfreeman, I appreciate your contribution! I'll have a look in the next few days.

**Comment by medfreeman** [comment]
Hey, what do you think of my PR ?
Thanks

**Comment by tyxla** [comment]
Hey @medfreeman, thanks for your PR!

To be honest, I'm not anxious to introduce external JS files into the plugin, especially if they include functions from the original plugin. That wouldn't be a problem if Gravity Forms were open source, but unfortunately that's not the case.

**Comment by medfreeman** [comment]
Gravity Forms is indeed opensource, GPL-3.0.
And this is about a single function.
I can add the missing credits if you'll reconsider this.
Thanks

**Comment by tyxla** [comment]
@medfreeman good call, I didn't expect for them to be GPL while being a premium plugin.
But it makes sense to use the code freely in that case, but indeed, let's add credits accordingly.

Can I recommend that you make the code compliant with the WordPress Coding Standards?
Here are some links for more information:

* [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)
* [WordPress JavaScript Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/)

Thanks in advance! :+1:
